index.d.ts

```diff
@@ -0,0 +1,101 @@
+export interface Action<T> {
+  type: T;
+}
+
+
+/* reducers */
+
+export type Reducer<S> = <A extends Action<any>>(state: S, action: A) => S;
+
+export interface ReducersMapObject {
+  [key: string]: Reducer<any>;
+}
+
+export function combineReducers<S>(reducers: ReducersMapObject): Reducer<S>;
+export function combineReducers<S, M extends ReducersMapObject>(
+  reducers: M
+): Reducer<S>;
+
+
+/* store */
+
+export interface Dispatch {
+  <A>(action: A): A;
+  <A, B>(action: A): B;
+}
+
+export interface Unsubscribe {
+  (): void;
+}
+
+export interface Store<S> {
+  dispatch: Dispatch;
+  getState(): S;
+  subscribe(listener: () => void): Unsubscribe;
+  replaceReducer(reducer: Reducer<S>): void;
+}
+
+export interface StoreCreator {
+  <S>(reducer: Reducer<S>): Store<S>;
+  <S>(reducer: Reducer<S>, initialState: S): Store<S>;
+
+  <S>(reducer: Reducer<S>, enhancer: StoreEnhancer): Store<S>;
+  <S>(reducer: Reducer<S>, initialState: S, enhancer: StoreEnhancer): Store<S>;
+}
+
+export type StoreEnhancer = (next: StoreCreator) => StoreCreator;
+
+export const createStore: StoreCreator;
+
+
+/* middleware */
+
+export interface MiddlewareAPI<S> {
+  dispatch: Dispatch;
+  getState(): S;
+}
+
+export interface Middleware {
+  <S>(api: MiddlewareAPI<S>): (next: Dispatch) => <A, B>(action: A) => B;
+}
+
+export function applyMiddleware(...middlewares: Middleware[]): StoreEnhancer;
+
+
+/* action creators */
+
+export interface ActionCreator<A> {
+  (...args: any[]): A;
+}
+
+export interface ActionCreatorsMapObject {
+  [key: string]: ActionCreator<any>;
+}
+
+
+export function bindActionCreators<A extends ActionCreator<any>>(
+  actionCreator: A, dispatch: Dispatch
+): A;
```

Comment:
Actually there are no restriction on shape of action creator. Our proposition is just reflection of real world. But you of course can lock in safer pattern inside your project, just:

```typescript
interface MyActionCreator<A> extends ActionCreator<A> {
  (x: A): A;
}

const square: MyActionCreator<number> = (x: number) => x * x;
const x: number = square(1);
const y: number = square('bad'); // unsafe but works
```

but of course in this situation it would be much better to let typescript infer types properly:

```typescript
const square = (x: number) => x * x; // inferred as    (x: number) => number
const x = square(1); // safe and compiles, x is inferred as number
const y = square('bad'); // unsafe and breaks compilation
```